mindmeld/auto_annotator.py

```diff
@@ -835,5 +839,300 @@ def valid_entity_check(self, entity):
+    """Custom Annotator class used to generate annotations."""
+
+    def __init__(self, app_path, config=None):
+        """Initializes an annotator.
+
+        Args:
+            app_path (str): The location of the MindMeld app
+            config (dict, optional): A config object to use. This will
+                override the config specified by the app's config.py file.
+        """
+        super().__init__(app_path=app_path, config=config)
+        self.language = self.config.get("language", "en")
+        self.translator = self.get_translator()
+        self.en_annotator = SpacyAnnotator(app_path)
+
+    def get_translator(self):
+        """Creates a Translator object based on the translator class specified in the config.
+        If a translator is not specified, Duckling heuristics will be used.
+
+        Returns:
+            translator (Translator): Translator specified in the config, None if not specified.
+        """
+        if self.config.get("translator"):
+            translator_class_name = self.config.get("translator", "GoogleTranslator")
+            if translator_class_name == "GoogleTranslator":
+                return GoogleTranslator()
+            else:
+                logger.warning(
+                    "%s is not a valid translation class. Currently only ‘GoogleTranslator’"
+                    " is supported.", translator_class_name
+                )
+        logger.info(
+            "No Translator selected. MLA will detect entities based on Duckling heuristics."
+        )
+
+    def parse(self, sentence, entity_types=None, language=None, **kwargs):
+        """
+        Args:
+            sentence (str): Sentence to detect entities.
+            entity_types (list): List of entity types to parse. If None, all
+                    possible entity types will be parsed.
+            language (str): Language code for the language of the given sentence
+        Returns: entities (list): List of entity dictionaries.
+        """
+        language = language or self.language
+        params = {"sentence": sentence, "language": language}
+        duckling_entities = (
+            self._parse_with_translator(**params)
+            if self.translator
+            else self._parse_without_translator(**params)
+        )
+        if entity_types:
+            duckling_entities = [
+                e for e in duckling_entities if e["dim"] in entity_types
+            ]
+        return MultiLingualAnnotator.convert_duckling_candidates(duckling_entities)
+
+    def _parse_with_translator(self, sentence, language=None):
+        """Parse helper function to be used if a translator is present.
+
+        Args:
+            sentence (str): Sentence to detect entities.
+            language (str): Language code for the language of the given sentence
+        Returns: entities (list): List of duckling candidates.
+        """
+        candidates = self.en_annotator.duckling.get_candidates_for_text(
```

Comment:
I think the candidates returned from `get_candidates_for_text ` should be a list of `Spans` instead of `tuples`. That way, the functions like `_has_overlap ` and `_is_larger` can be neatly placed in the `Span` class.

Reply:
Added `has_overlap(object)` to the span class. is_larger() is no longer needed since functions like `__gt__` are present